Add support for match_lists and using them in logging policy

quulah opened the pull request:
## Change Summary

<!-- Enter short PR description -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Related to #979

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
<!--- Describe your changes in detail -->
<!--- Describe data model implemented for new features -->

- Add supports for specifying `match-list`s and using them with the `logging policy`

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code has been rebased from devel before I start
- [X] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contributing/) document.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [X] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

Commit:
Add support for match_lists and using them in logging policy

Commit:
Molecule tests for match-list

quulah commented:
Rebased on current devel and pushed the Molecule tests for this.

I have added a check to the `logging` part that checks whether the `match-list` exists. I'm second guessing my decision here, any thoughts on that?

Might be simpler without it and you could add `match-list` via other configlets potentially.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/logging.j2:
```diff
@@ -43,4 +43,11 @@ logging host {{ host }}
 {%             endfor %}
 {%         endif %}
 {%     endfor %}
+{%     if logging.policy.match.match_lists is arista.avd.defined %}
```
This test is not required, since `arista.avd.natural_sort` has it built-in. So you can safely remove it.

quulah replied:
Fixed in 3e2045270963f2545c39be7586a240eff81edffe

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/logging.j2:
```diff
@@ -43,4 +43,11 @@ logging host {{ host }}
 {%     endfor %}
+{%     if logging.policy.match.match_lists is arista.avd.defined %}
+{%         for match_list in logging.policy.match.match_lists | arista.avd.natural_sort %}
+{%             if match_list in match_lists | arista.avd.natural_sort %}
```
I understand your reasoning for this, but `eos_cli_config_gen` should not do any "clever" evaluation of the configuration. Just output the CLI if you have the variables needed. (You where already mentioning this, so I hope you agree with this background).

quulah replied:
Yes, I totally agree. :)

quulah replied:
Fixed in 3e2045270963f2545c39be7586a240eff81edffe

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -1347,6 +1358,11 @@ logging:
+    match:
+      match_lists:
+        < match_list >:
+          action: < action >
```
please provide syntax with valid actions to help the user. (`discard` is the only valid option for now)

quulah replied:
Fixed in 16b86309c7f14f912fcdff03f5792d471ad057e6

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
````diff
@@ -544,6 +545,16 @@ route_maps:
+#### Match Lists
+
+```yaml
+match_lists:
````
EOS supports `match-list input prefix-ipv4 <name>` and `match-list input string <name>`. 
Example:
```eos
match-list input string test
   10 match regex 10
match-list input prefix-ipv4 test2
   match prefix-ipv4 2.2.2.2/32
   match prefix-ipv4 1.1.1.1/32
```
We should make sure that the data model is close to the cli model, so it can be extended to cover everything we have in EOS. Since the syntax varies a lot between the two match-list types, we need to build seperate data models.
I suggest a few minor changes:
```yaml
match_list:
  input:
    string:
      < match_list_1 >:
        sequence_numbers:
          < sequence_id 1 >:
            match_regex: < match string >
```
Alternatively start with `match_list_input` since `input` is the only option after `match-list`

quulah replied:
I think a shallower dictionary is probably better, so would go with `match_list_input`.

Do you want me to add support for `prefix-ipv4` too? Should be pretty straightforward, if the parameters always look like the ones in your example.

ClausHolbechArista replied:
OK go with `match_list_input`
Re `prefix-ipv4` it is fine if you just add what you need and know the CLI for.

quulah replied:
Changed to `match_list_input` in 61f6cfa40d352d108cc6174a60d9eb02229dd482.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/match-lists.j2:
```diff
@@ -0,0 +1,10 @@
+{%     for match_list in match_lists | arista.avd.natural_sort %}
+match-list input string {{ match_list }}
+{%         for sequence in match_lists[match_list].sequence_numbers | arista.avd.natural_sort %}
+   {{ sequence }} match {{ match_lists[match_list].sequence_numbers[sequence].match }}
```
missing a test for `match_lists[match_list].sequence_numbers[sequence].match`

quulah replied:
Fixed in d045a8f9c6df50b2f372a593e770eae35a65b9d2

ClausHolbechArista reviewed (changes requested):
A few requested changes. Let us agree on data model before you change anything

Commit:
Remove unnecessary checks from template

Commit:
Use valid action in documentation

Commit:
Ensure match is defined

Commit:
Change match-list input string variable to match commandline

Commit:
Add generated documentation for match-lists

quulah commented:
> A few requested changes. Let us agree on data model before you change anything

All of these should now be done. I also rebased to current devel and added the generated documentation for `match_list_input`.

ClausHolbechArista reviewed (approved):
LGTM! Thank you for this contribution!
